zcash_primitives/src/zip316/transparent.rs

```diff
@@ -0,0 +1,99 @@
+use crate::sapling::keys::{prf_expand_vec, OutgoingViewingKey as SaplingOvk};
+use crate::zip32::ChainCode;
+use std::convert::TryInto;
+use zcash_address::unified::{Container, Fvk, Ufvk};
+
+/// The transparent full viewing key as specified in [ZIP 316][transparent-fvk].
+///
+/// [transparent-fvk]: https://zips.z.cash/zip-0316#encoding-of-unified-full-incoming-viewing-keys
+pub struct FullViewingKey {
+    c: ChainCode,
+    pk: secp256k1::PublicKey,
+}
+
+/// Internal ovk used for autoshielding.
+pub struct InternalOvk([u8; 32]);
+
+impl InternalOvk {
+    pub fn as_bytes(&self) -> [u8; 32] {
+        self.0
+    }
+}
+
+/// External ovk used by zcashd for transparent -> shielded spends to
+/// external receivers.
+pub struct ExternalOvk([u8; 32]);
+
+impl ExternalOvk {
+    pub fn as_bytes(&self) -> [u8; 32] {
+        self.0
+    }
+}
+
+impl FullViewingKey {
+    /// Derives the internal ovk and external ovk corresponding to this
+    /// transparent fvk. As specified in [ZIP 316][transparent-ovk].
+    ///
+    /// [transparent-ovk]: https://zips.z.cash/zip-0316#deriving-internal-keys
+    fn ovk_for_shielding(&self) -> (InternalOvk, ExternalOvk) {
+        let i_ovk = prf_expand_vec(&self.c.as_bytes(), &[&[0xd0], &self.pk.serialize()]);
+        let i_ovk = i_ovk.as_bytes();
+        let ovk_internal = InternalOvk(i_ovk[..32].try_into().unwrap());
+        let ovk_external = ExternalOvk(i_ovk[32..].try_into().unwrap());
+
+        (ovk_internal, ovk_external)
+    }
+
+    /// Derives the internal ovk corresponding to this transparent fvk.
+    pub fn internal_ovk(&self) -> InternalOvk {
+        self.ovk_for_shielding().0
+    }
+
+    /// Derives the external ovk corresponding to this transparent fvk.
+    pub fn external_ovk(&self) -> ExternalOvk {
+        self.ovk_for_shielding().1
+    }
+}
+
+impl From<(ChainCode, secp256k1::PublicKey)> for FullViewingKey {
+    fn from(c_pk: (ChainCode, secp256k1::PublicKey)) -> FullViewingKey {
+        FullViewingKey {
+            c: c_pk.0,
+            pk: c_pk.1,
+        }
+    }
+}
+
+impl std::convert::TryFrom<[u8; 65]> for FullViewingKey {
+    type Error = secp256k1::Error;
+    fn try_from(bytes: [u8; 65]) -> Result<Self, Self::Error> {
+        let c: [u8; 32] = bytes[..32].try_into().unwrap();
+        let c: ChainCode = c.into();
+        let pk = secp256k1::PublicKey::from_slice(&bytes[32..])?;
+        Ok((c, pk).into())
+    }
+}
+
+/// A transparent internal ovk can be used to shield to Sapling.
+impl From<InternalOvk> for SaplingOvk {
+    fn from(t_int_ovk: InternalOvk) -> SaplingOvk {
+        SaplingOvk(t_int_ovk.0)
+    }
+}
+
+/// Finds the internal or external ovk for the transparent component
+/// of a given Ufvk, if it exists.
+pub fn find_ovk(key: &Ufvk, external: bool) -> Option<[u8; 32]> {
+    for r in &key.items() {
+        if let Fvk::P2pkh(data) = r {
+            let fvk: FullViewingKey = (*data).try_into().ok()?;
```

Comment:
This is similar but different to https://github.com/zcash/zcash/pull/5470/files#diff-98a0abefb7ae486bd76f5e8d28d7fd65cead55c7344b31cc90caa56583917ca3R155-R159 . Why the difference?

Reply:
Oh it's just because the `TryInto` implementation above fills in the gap.